src/Resources/ResourceManager/Extensions/JTokenExtensions.cs

```diff
@@ -123,5 +123,91 @@ internal static object ConvertPropertyValueForPsObject(JToken propertyValue)
 
             return propertyValue.ToString();
         }
+
+        /// <summary>
+        /// Deep converts a <see cref="JObject"/> to <see cref="PSObject"/>
+        /// </summary>
+        /// <param name="jtoken">The <see cref="JObject"/></param>
+        /// <param name="objectType">The type of the object.</param>
+        internal static PSObject ToPsObjectByDeepConvert(this JToken jtoken, string objectType = null)
+        {
+            if (jtoken == null)
+            {
+                return null;
+            }
+
+            if (jtoken.Type != JTokenType.Object)
+            {
+                return new PSObject(JTokenExtensions.ConvertPropertyValueForPsObject(propertyValue: jtoken));
```

Comment:
This branch isn't doing deep convert. For example if you pass this object to ToPsObjectByDeepConvert():
```
[
    {
        "obj1": {
            "val1": "abc",
            "obj2": {
                 "obj3": {
                    "val2": "def",
                    "val3": "ghi",
                    "arr1": [
                        "val4", "val5", "val6"
                    ]
                }
            }
        }
    },
    {
        "val5": "jkl"
    }
]
```

... it will call ToPSObject() on each element of the array, which will not do a deep convert on obj1, which means it will not convert below obj2 level (as you observed), which will skip obj3 and arr1 (turn them into strings).

Reply:
This is a typo. I can fix it but I'll go with your suggestion of eliminating code duplication in a later comment. 

---
In reply to: [414313776](https://github.com/Azure/azure-powershell/pull/11644#discussion_r414313776) [](ancestors = 414313776)